telemetry/jetbrains/src/main/kotlin/software/aws/toolkits/telemetry/generator/GenerationFunctions.kt

```diff
@@ -0,0 +1,164 @@
+
+    item.allowedValues!!.forEach { enumValue ->
+        enum.addEnumConstant(
+            enumValue.toString().toUpperCase().filterInvalidCharacters(), TypeSpec.anonymousClassBuilder()
+                .addSuperclassConstructorParameter("%S", enumValue.toString())
+                .build()
+        )
+    }
+
+    val companion = TypeSpec.companionObjectBuilder()
+        .addFunction(
+            FunSpec.builder("from")
+                .returns(ClassName("", item.name.toTypeFormat()))
+                .addParameter("type", Any::class)
+                .addStatement("return values().first { it.value == type.toString() }")
+                .build()
+        )
+        .build()
+
+    enum.addType(companion)
+
+    addType(enum.build())
+}
+
+fun FileSpec.Builder.generateTelemetryObjects(item: TelemetryDefinition) =
+    item
+        .metrics
+        .sortedBy { it.name }
+        .groupBy { it.name.split("_").first().toLowerCase() }
+        .forEach { metrics: Map.Entry<String, List<Metric>> -> generateNamespaces(item.types!!, metrics.key, metrics.value) }
+
+fun FileSpec.Builder.generateNamespaces(types: List<TelemetryMetricType>, namespaceType: String, metrics: List<Metric>) {
+    val namespace = TypeSpec.objectBuilder("${namespaceType.toTypeFormat()}Telemetry")
+    metrics.forEach { namespace.generateRecordFunctions(it, types) }
+    addType(namespace.build())
+}
+
+fun TypeSpec.Builder.generateRecordFunctions(metric: Metric, types: List<TelemetryMetricType>) {
+    // metric.name.split("_")[1] is guaranteed to work at this point because the schema requires the metric name to have at least 1 underscore
+    val functionName = metric.name.split("_")[1]
+    val parameters = buildParameters(metric, types)
+    val functionBuilder = FunSpec.builder(functionName)
+    functionBuilder.addParameters(parameters)
+    functionBuilder.generateFunctionBody(metric)
+    functionBuilder.addKdoc(metric.description)
+    val function = functionBuilder.build()
+    addFunction(function)
+    // Result is special cased to generate a function that accepts true/false instead of a Result
+    if (metric.metadata?.any { it.type == RESULT } != true) {
+        return
+    }
+    val resultOverloadFunction = FunSpec.builder(functionName)
+    val resultOverloadParameters = parameters.map {
+        if (it.name == RESULT) {
+            ParameterSpec.builder(SUCCESS, BOOLEAN).build()
+        } else {
+            it
+        }
+    }
+    resultOverloadFunction.addParameters(resultOverloadParameters)
+    resultOverloadFunction.generateResultOverloadFunctionBody(function, resultOverloadParameters)
+    resultOverloadFunction.addKdoc(metric.description)
+    addFunction(resultOverloadFunction.build())
+}
+
+fun buildParameters(metric: Metric, types: List<TelemetryMetricType>): List<ParameterSpec> {
+    val projectParameter = ClassName("com.intellij.openapi.project", "Project").copy(nullable = true)
+    val list = mutableListOf<ParameterSpec>()
```

Comment:
If only the collection builders were out...

Reply:
yep, tried to fix this but settled on this. collection builders would be nice.